compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -816,8 +816,13 @@ impl<'ast, 'ra, 'tcx> LateResolutionVisitor<'_, 'ast, 'ra, 'tcx> {
                 return (true, suggested_candidates, candidates);
             }
 
-            // If the first argument in call is `self` suggest calling a method.
-            if let Some((call_span, args_span)) = self.call_has_self_arg(source) {
+            // If the first argument in call is `self` suggest calling a method,
+            // even if we didn't find an associated item. However, only suggest
+            // this if there's no `Res`, since if there was a resolution but it
+            // was invalid, it's more likely not a typo of this form.
+            if let Some((call_span, args_span)) = self.call_has_self_arg(source)
+                && res.is_none()
```

Comment:
This doesn't affect any *other* tests, so I just went with a simple heuristic here.

We could alternatively check if the path segment starts with a capitalized identifier, or match `!matches!(res, Some(DefKind::Struct | DefKind::Variant))`, or something? I think that's kinda overkill tho.